NOnion/TorCircuit.fs

```diff
@@ -0,0 +1,227 @@
+
+                let recognized =
+                    System.BitConverter.ToUInt16 (
+                        decryptedRelayCellBytes,
+                        Constants.RelayRecognizedOffset
+                    )
+
+                if recognized <> Constants.RecognizedDefaultValue then
+                    failwith "Incoming cell is unrecognizable"
+
+                let digest =
+                    decryptedRelayCellBytes
+                    |> Array.skip Constants.RelayDigestOffset
+                    |> Array.take Constants.RelayDigestLength
+
+                Array.fill
+                    decryptedRelayCellBytes
+                    Constants.RelayDigestOffset
+                    Constants.RelayDigestLength
+                    0uy
+
+                let computedDigest =
+                    cryptoState.BackwardDigest.PeekDigest
+                        decryptedRelayCellBytes
+                        0
+                        decryptedRelayCellBytes.Length
+                    |> Array.take Constants.RelayDigestLength
+
+                if digest <> computedDigest then
+                    failwith "Digest verification for incoming cell failed"
+
+                cryptoState.BackwardDigest.Update
+                    decryptedRelayCellBytes
+                    0
+                    decryptedRelayCellBytes.Length
+
+                let decryptedRelayCell =
+                    CellPlainRelay.FromBytes decryptedRelayCellBytes
+
+
+                (decryptedRelayCell.StreamId, decryptedRelayCell.Data)
+            | _ ->
+
+                failwith "Unexpected state when receiving relay cell"
+
+        RunSyncWithSemaphore controlLock safeDecryptCell
+
+    member self.CreateFast () =
+        async {
+            let createFast () =
+                async {
+                    let circuitId = guard.RegisterCircuit self
+
+                    let randomClientMaterial =
+                        Array.zeroCreate<byte> Constants.HashLength
+
+                    RandomNumberGenerator
+                        .Create()
+                        .GetBytes randomClientMaterial
+
+                    let completionTcs = TaskCompletionSource ()
+
+                    circuitState <-
+                        CreatingFast (
+                            circuitId,
+                            randomClientMaterial,
+                            completionTcs
+                        )
+
+                    do!
+                        guard.Send
+                            circuitId
+                            {
+                                CellCreateFast.X = randomClientMaterial
+                            }
+
+                    return completionTcs.Task
+                }
+
+            let! completionTask = RunAsyncWithSemaphore controlLock createFast
+
+            //FIXME: Connect Timeout?
+            return! completionTask |> Async.AwaitTask
+        }
+
+    member self.CreateFastAsTask () =
```

Comment:
my bad, actually the suffix should be `Async`, not `AsTask`, to match with C# API guidelines

Reply:
^

Reply:
Fixed